### PR TITLE
Add Salmon quantification tool

### PR DESCRIPTION
**Pull request by jkokosar**
- [x] Requires resolwebio/rnaseq:4.5.0 Docker

### CODE REVIEW

**Comment by jkokosar** [comment]
@romunov please review, especially the tximport_summarize.R part

File: resolwe_bio/processes/expression/salmon_index.yml [hunk 1]
```diff
@@ -86,9 +117,12 @@
 
       salmon --no-version-check index \
         -t {{ nucl.fastagz.file }} \
+        {% if decoys %} -d {{ decoys.file.file }} {% endif %} \
+        {% if gencode %} --gencode {% endif %} \
+        {% if keep_duplicates %} --keepDuplicates {% endif %} \
+        {% if perfect_hash %} --perfectHash {% endif %} \
         -i salmon_index \
         -p {{ requirements.resources.cores }} \
-        --type quasi \
```
Comment (romunov) [inline resolwe_bio/processes/expression/salmon_index.yml]: Was this argument deprecated?

File: resolwe_bio/tools/tximport_summarize.R [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env Rscript
+library(argparse)
+library(GenomicFeatures)
+library(readr)
```
Comment (romunov) [inline resolwe_bio/tools/tximport_summarize.R]: Can't find where functions from this package are used.

Comment (jkokosar) [reply resolwe_bio/tools/tximport_summarize.R]: True, will remove.

File: resolwe_bio/processes/expression/salmon_index.yml (hunk 1 above)
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon_index.yml]: The `quasi` is the only (and default) option for `--type` in recent version of Salmon, which removes the need to have this parameter exposed.

File: resolwe_bio/processes/expression/salmon.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+                            "larger values imply a more fine-grained "
+                            "factorization.  If range factorization "
+                            "is enabled, a common value to select "
+                            "for this parameter is 4."
+            )
+
+            min_assigned_frag = IntegerField(
+                label="--minAssignedFrags",
+                default=10,
+            )
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Will add :)

File: resolwe_bio/processes/expression/salmon.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                label="--noLengthCorrection",
+                default=False,
+                description="[Experimental] Entirely disables "
+                            "length correction when estimating the "
+                            "abundance of transcripts.  This option "
+                            "can be used with protocols where one "
+                            "expects that fragments derive from "
+                            "their underlying targets without regard "
+                            "to that target's length (e.g. QuantSeq).",
+            )
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Added

File: resolwe_bio/processes/expression/salmon.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+        if inputs.salmon_index.build != inputs.annotation.build:
+            self.error('Salmon index file build ({}) must match GTF annotation '
+                       'file build ({})'.format(inputs.salmon_index.build, inputs.annotation.build))
+
+        if inputs.salmon_index.source != inputs.annotation.source:
+            self.error('Salmon index file source ({}) must match GTF annotation '
+                       'file source ({})'.format(inputs.salmon_index.source, inputs.annotation.source))
+
+        if inputs.options.gc_bias and inputs.options.no_length_correction:
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Done, good suggestion.

File: resolwe_bio/processes/expression/salmon.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        if inputs.options.validate_mappings and inputs.options.min_score_fraction > 0:
+            args.extend(['--minScoreFraction', inputs.options.min_score_fraction])
+
+        if inputs.options.consensus_slack is not None:
+            args.extend(['--consensusSlack', inputs.options.consensus_slack])
+
+        # Run Salmon Quant
+        return_code, _, _ = Cmd['salmon']['quant'][args] & TEE(retcode=None)
+        if return_code:
+            self.error("Error while running Salmon Quant.")
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Not sure if we can do that. Will skip for now.

File: resolwe_bio/processes/expression/salmon.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        # Save the abundance estimates to JSON storage
+        Cmd['expression2storage.py']('--output', 'json.txt', reads_name + output_suffix)
+
+        # Prepare expression set file with feature_id -> gene_id mappings
+        exp_set_args = [
+            '--expressions', reads_name + output_suffix,
+            '--source_db', inputs.salmon_index.source,
+            '--species', inputs.salmon_index.species,
+            '--output_name', reads_name + '_expressions',
+            '--expressions_type', 'TPM',
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Fixed, thanks!

File: resolwe_bio/tools/deseq.R [hunk 8]
```diff
@@ -31,6 +32,10 @@ if (args$format == 'rsem') {
     txi <- tximport(files, type='rsem')
     rownames(sampleTable) <- colnames(txi$counts)
     dds <- tryCatch(DESeqDataSetFromTximport(txi, sampleTable, ~condition), error=error)
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/tools/deseq.R]: @mzganec designed this to print out the process error: https://github.com/genialis/resolwe-bio/blob/729acb04a96b68ebecc8acfbceee00741221cdf6/resolwe_bio/tools/deseq.R#L8 Is the above proposed solution still needed or better in any way?

[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/tools/deseq.R]: I missed this detail. It makes more sense now, thanks.

File: resolwe_bio/tests/processes/test_expression.py [hunk 9]
```diff
@@ -554,6 +559,56 @@ def test_salmon_index(self):
...
+                'consensus_slack': 0.25,
+                'no_length_correction': False,
+                'discard_orphans_quasi': True,
+            }
+        }
+        salmon_quant = self.run_process('salmon-quant', inputs)
+        self.assertFile(
+            salmon_quant,
+            'exp_set',
+            os.path.join('salmon_quant/output', 'salmon_quant_tpm.tab.gz'),
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_expression.py]: `os.path.join('salmon_quant', 'output', 'salmon_quant_tpm.tab.gz'),`

File: resolwe_bio/tools/tximport_summarize.R [hunk 10]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env Rscript
+library(argparse)
+library(GenomicFeatures)
+library(tximport)
+
+parser = ArgumentParser(description='Summarize TPM values to gene-level expressions.')
+parser$add_argument('infile', help="Salmon output file (transcript-level).")
+parser$add_argument('outfile', help='GTF file used for transcript-to-gene mapping.')
```
Comment (mzganec) [inline resolwe_bio/tools/tximport_summarize.R]: Incorrect help string.

File: resolwe_bio/tools/tximport_summarize.R [hunk 11]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env Rscript
+library(argparse)
+library(GenomicFeatures)
+library(tximport)
+
+parser = ArgumentParser(description='Summarize TPM values to gene-level expressions.')
```
Comment (mzganec) [inline resolwe_bio/tools/tximport_summarize.R]: It's usually a good choice to list all output related arguments after all input arguments.

File: resolwe_bio/tools/tximport_summarize.R [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+parser$add_argument('outfile', help='GTF file used for transcript-to-gene mapping.')
+parser$add_argument('gtf', help='GTF file used for transcript-to-gene mapping.')
+args = parser$parse_args(commandArgs(trailingOnly=TRUE))
+
+# Create transcript-to-gene mapping table
+db_from_gtf = makeTxDbFromGFF(args$gtf)
+k = keys(db_from_gtf, keytype = "TXNAME")
+tx2gene = select(db_from_gtf, k, "GENEID", "TXNAME")
+# Save transcript-to-gene mapping table to file
+write.table(tx2gene, file='tx2gene.txt', sep="\t", quote = FALSE, row.names = FALSE)
```
Comment (mzganec) [inline resolwe_bio/tools/tximport_summarize.R]: The output file name of this table should be a command-line parameter. In the process, `tx2gene` file name should be derived from the name of the annotation file.

File: resolwe_bio/processes/expression/salmon.py [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+        if inputs.reads.type.startswith('data:reads:fastq:single:'):
+            args.extend(['-r'] + [lane.path for lane in inputs.reads.fastq])
+        else:
+            args.extend(['-1'] + [lane.path for lane in inputs.reads.fastq])
+            args.extend(['-2'] + [lane.path for lane in inputs.reads.fastq2])
+
+        # Prepare optional inputs
+        if inputs.options.seq_bias and not inputs.options.no_length_correction:
+            args.append('--seqBias')
+        else:
```
Comment (mzganec) [inline resolwe_bio/processes/expression/salmon.py]: I'm not sure that the condition to trigger this warning is correct.

See the following truth table. Discrepancy occurs in the last two rows.

|inputs.options.seq_bias|inputs.options.no_length_correction|display warning (current impl.)|display warning (desired impl.)|
|:-:|:-:|:-:|:-:|
|True|True|True|True|
|True|False|False|False|
|False|True|True|False|
|False|False|True|False|

The same applies to the next warning as well.

Comment (jkokosar) [reply resolwe_bio/processes/expression/salmon.py]: Fixed in https://github.com/genialis/resolwe-bio/pull/794

**Pull request merged**